### ISSUE TITLE
Support images hosted in ECR for RKE, custom, and node-driver clusters

### ISSUE THREAD

**Issue by dkeightley** [rancher/rancher#28693]
**What kind of request is this (question/bug/enhancement/feature request):** feature request

**Steps to reproduce (least amount of steps as possible):**

- Attach an instance profile [with the appropriate instance profile permissions](https://docs.aws.amazon.com/AmazonECR/latest/userguide/ECR_on_EKS.html) to each node in the cluster

- [Populate ECR with rancher-images.txt](https://rancher.com/docs/rancher/v2.x/en/installation/other-installation-methods/air-gap/populate-private-registry/#3-save-the-images-to-your-workstation) to simulate a private registry / air gap for Rancher container images. 
  - **Note 1** the location where the script is run must have adequate IAM permissions (instance profile, or access/secret keys) and [authenticated with ECR](https://docs.aws.amazon.com/AmazonECR/latest/userguide/getting-started-cli.html#cli-authenticate-registry), or [using the ecr-credential-helper](https://github.com/awslabs/amazon-ecr-credential-helper).
  - **Note 2** repos are not auto-created in ECR on push, loop through `aws ecr create-repository --repository-name <repo name>` to create these first.

- Update the [default registry, or add a private_registry per-cluster](https://rancher.com/docs/rancher/v2.x/en/admin-settings/config-private-registry/).

- ECR uses a token with 12h expiry, therefore a user/pass would not be possible/ideal option (frequent automation would be required for the default/private_registry).

**Result:**

Kubelet will auth and pull images for pods, local `docker` commands on the node will succeed (after auth to ECR, or when using the ecr-credential-helper)

The docker go HTTP client making API calls via the websocket is unable to benefit from the nodes `config.json` so RKE-based actions will fail to pull images (upgrades, reconciles, snapshots?)

gz#12074

**Comment by slickwarren** [comment]
waiting for a 2.6 RC to tag images + add to ECR

**Comment by deniseschannon** [comment]
Available to test in v2.6.0-rc1

**Comment by slickwarren** [comment]
using v2.6.0-rc1, I was able to bring up an ECR registry and create an RKE1 cluster using v1.3.0-rc10. 
```
services:
  kubelet:
    extra_env:
      - "AWS_ACCESS_KEY_ID=<access_id>"
      - "AWS_SECRET_ACCESS_KEY=<secret_key>"
private_registries:
     - url: <account_ID>.dkr.ecr.us-west-2.amazonaws.com
       is_default: true
       credentialPlugin:
        type: ecr
```

However, I tried adding this snippet to the yaml during provisioning of an ec2 cluster through rancher, and was unable to complete provisioning. I got this error
```

Unable to find image '<account_ID>.dkr.ecr.us-west-2.amazonaws.com/rancher/rancher-agent:v2.6.0-rc1' locally docker: Error response from daemon: Head https://<account_ID>.dkr.ecr.us-west-2.amazonaws.com/v2/rancher/rancher-agent/manifests/v2.6.0-rc1: no basic auth credentials. See 'docker run --help'. Process exited with status 125 : exit status 1

```
I get a similar error when using a `custom` cluster related to basic auth credentials. Could someone provide some instructions here? Based on the above config, it is not clear what should be done in this case

**Comment by ibrokethecloud** [comment]
For using ecr as the default registry we need to pass the iam role credentials to the plugin as well.

```yaml
services:
  kubelet:
    extra_env:
      - "AWS_ACCESS_KEY_ID=your_access_key"
      - "AWS_SECRET_ACCESS_KEY=your_secret_key"         
private_registries:
     - url: account-id.dkr.ecr.region.amazonaws.com
       is_default: true
       credentialPlugin:
        type: ecr
        # Static Credentials
        #aws_access_key_id: your_access_key
        #aws_secret_access_key: your_secret_key
        #aws_session_token: your_session_token
```

Since at this point k8s is not up we need to send the credentials to docker to allow it to authenticate and pull the infra images.

**Comment by slickwarren** [comment]
using a custom cluster on v2.6.0-rc1, I was still unable to successfully get an ECR cluster to active. I did add the 
```
aws_access_key_id: your_access_key
aws_secret_access_key: your_secret_key
```
to my yaml and have installed + configured the[ ecr credential helper](https://github.com/awslabs/amazon-ecr-credential-helper) on the downstream node. The node is able to register, however part of the setup fails, with this error 
```
[Failed to pull image [<account_id>.dkr.ecr.us-west-2.amazonaws.com/rancher/rke-tools:v0.1.78] on host [<ip>]: NoCredentialProviders: no valid providers in chain. Deprecated. For verbose messaging see aws.Config.CredentialsChainVerboseErrors]
```
I've double checked that image does exist in my ECR setup.

**Comment by ibrokethecloud** [comment]
@slickwarren some changes have been pushed in another PR.

New format for passing creds would be as follows:

```yaml:
services:
  kubelet:
    extra_env:
      - "AWS_ACCESS_KEY_ID=ACCESSKEY"
      - "AWS_SECRET_ACCESS_KEY=SECRETKEY"
private_registries:
     - url: ACCOUNTID.dkr.ecr.ap-southeast-2.amazonaws.com
       is_default: true
       ecrCredentialPlugin: 
        aws_access_key_id: "ACCESSKEY"
        aws_secret_access_key: "SECRETKEY"
```

I have tested this with a custom rancher build, and now rancher is able to pull images just fine from ECR for downstream cluster provisioning.

**Comment by kinarashah** [comment]
Available to test with v2.6.0-rc5.

**Comment by slickwarren** [comment]
on 2.6.0-rc5:
* add custom cluster using ECR registry (node must have ecr helper installed) -- pass
* upgrade cluster k8s version after 12 hours -- pass
* deploy workload from ECR after 12 hours -- pass
